### ISSUE TITLE
[Selenium 4] [C#] RemoteWebDriver doesn't implement IDevTools

### ISSUE THREAD

**Issue by szamacz** [SeleniumHQ/selenium#9017]
## Description

When trying to get devtools for RemoteWebdriver as IDevTools, null returned.
When instead of RemoteWebdriver using ChromeDriver, everything works fine and able to create dev tools session.
Looks that RemoteWebDriver doesn't implement IDevTools. [https://github.com/SeleniumHQ/selenium/blob/trunk/dotnet/src/webdriver/Remote/RemoteWebDriver.cs](url)

## To Reproduce

Doesn't work - devTools are null
`
var driver = new RemoteWebDriver(new Uri(_seleniumHubUrl), options.ToCapabilities(), TimeSpan.FromMinutes(2));
`
`var devTools = driver as IDevTools;`
`devTools.CreateDevToolsSession()`

Works fine - devTools are not null
`var driver = new ChromeDriver(Environment.CurrentDirectory, options, TimeSpan.FromMinutes(3));`
`var devTools = driver as IDevTools;`
`devTools.CreateDevToolsSession()`

## Expected behavior

RemoteWebDriver implements IDevTools.
Able to set up dev tools and create dev tools session for RemoteWebdriver

## Environment

OS: Windows10
Browser: Chrome
Browser version: 87
Browser Driver version: 87.0.4280.8800
Language Bindings version: C#  4.0.0-alpha07
Selenium Grid version (if applicable): 4.0.0-beta-1-prerelease-20201208

**Comment by SKumar-777** [comment]
Team,

I have tried to implement the devtools for RemoteWebdriver as IDevTools using C# and it returned null. It seems, the RemoteWebDriver does not implement IDevTools. We are waiting for this devtools support for RemoteWebdriver in Selenium 4.
Thanks in advance.

**Comment by szamacz** [comment]
I managed to have this working with some small workaround. Reviewing chromium driver code I found a method responsible for setting up chrome dev tools session and adjusted it a bit for my purpose.


First we need to set remote debugging address using chromeopitons.  What is also important we need to run our driver in headless mode.

```
options.AddArgument("--remote-debugging-port=9222");
options.AddArgument("--remote-debugging-address=0.0.0.0");  // this is needed as by default it is set to localhost which allows only local connections
options.AddArgument("headless");
```
Then it can be used for setting up remote WebDriver dev tools session

```
var chromeCapabilities = driver.Capabilities.GetCapability("goog:chromeOptions") as Dictionary<string, object>;
if (chromeCapabilities == null) throw new Exception("Cannot get chromeOptions capabilities");
var debuggerAddress = chromeCapabilities["debuggerAddress"].ToString();
```

```
 private static DevToolsSession SetupRemoteChromeDevToolsSession(string debuggerAddress)
        {
            var devToolsSession = new DevToolsSession(debuggerAddress);

            //without that we are getting deadlock from xunit due to async code running in synchronous matter
            var task = Task.Factory.StartNew(() => devToolsSession.Start().ConfigureAwait(false).GetAwaiter().GetResult()); 
            Task.WaitAll(task);
            return devToolsSession;
        }

```

Of course for our selenium docker image, we need to expose the proper port so chrome dev tools session can be set up, in my case, it was standalone chome

`docker run --privileged -d -p 4444:4444 -p 9222:9222 -v /dev/shm:/dev/shm selenium/standalone-chrome`

**Comment by SKumar-777** [comment]
Thanks szamacz! for your workaround and sample code. I could implement the devtools for RemoteWebdriver and able to download the browser console logs using Selenium 4 alpha -7. But this same workaround is not working in latest Selenium 4 **Beta1**. It seems few methods/properties were changed in latest Beta version. Please help us to achieve it.

Thanks in advance.

**Comment by Febbie15** [comment]
Team,

The above workaround is not working to implement Chrome DevTools support for Remote Webdriver in Selenium 4 **Beta2**.

We are waiting to have this native Chrome DevTools(IDevTools) support for Remote Webdriver in Selenium 4.

Thanks in advance.

**Comment by andreykavo** [comment]
Checked ChromiumDriver's sources, the code that initializes devTools is right there in the constructor: https://github.com/SeleniumHQ/selenium/blob/eaa10470f62dcfd04bdbfcf7775bcb32223d887d/java/client/src/org/openqa/selenium/chromium/ChromiumDriver.java#L88-L110
And this is a working example in Beta 1 (in Kotlin):
```
    private fun getDevtools(driver: RemoteWebDriver): DevTools {
        val factory = HttpClient.Factory.createDefault()
        val connection = ChromiumDevToolsLocator.getChromeConnector(
            factory,
            driver.capabilities,
            ChromeOptions.CAPABILITY
        )
        val cdpInfo = CdpVersionFinder().match(driver.capabilities.browserVersion)
            .orElseGet {
                throw Exception("Wrong version: ${driver.capabilities.browserVersion}")
            }

        return DevTools(cdpInfo::getDomains, connection.get())
    }

    fun main() {
        //...
        val devtools = getDevtools(driver)
        devtools.createSessionIfThereIsNotOne()
        devtools.send(Network.enable(Optional.empty(), Optional.empty(), Optional.empty()))
        //...
    }
```

**Comment by Febbie15** [comment]
Thanks andreykavo!

**Comment by Febbie15** [comment]
Thanks andreykavo! We are waiting to have this native Chrome DevTools(IDevTools) support for Remote Webdriver in Selenium 4 using **C#** . It seems few methods/properties were changed in latest Beta version. Please help us to achieve it.

**Comment by szamacz** [comment]
SKumar-777  Febbie15 My workaround won't work with Beta1 and Beta2 as guys decided to make method **Start** from **DevToolsSession** internal and it is not possible to set up devtools session for given debugger address anymore. On the other side, they didn't provide any way to connect to remote dev tools session. **andreykavo** thanks for posting workaround but in this case is a bit useless as the topic is not about java implementation.

**Comment by jimevans** [comment]
This has been fixed in 70af0d6
